### PR TITLE
Restrict using empty init for unbounded streams

### LINKED ISSUE #34966: When the completion type has no nil it should reflect in type param `CompletionType`
**Description:**
Looks like `next` methods return type is the union with `nil` and expected completion type. Changing the code to accept a `IntRecord|error?` from next removes the crash. But spec does not mention that the completion type can be `nil` implicitly. This could be a issue in type param analyser and should reflect in all the places we have `CompletionType`

#### Ballerina code
```ballerina
type IntRecord record {|
    int value;
|};

function init() {
    stream<int, error> emptyStream = new stream<int, error>();
    IntRecord|error returnedVal = emptyStream.next();
    if (returnedVal is IntRecord) {
        io:println(returnedVal.value);
    } else {
        io:println(returnedVal);
    }
}
```

#### Crash Log
```LogTalk
SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at s4.init(s4.bal:38)
        at $_init.$gen$$0046$0060init$0062(.:1)
        at $_init.$moduleInit(.)
        at $_init.$lambda$$moduleInit$(.)
        at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:611)
        at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:317)
        at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:284)
        at java.base/java.lang.Thread.run(Thread.java:834)
```
**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**
- https://github.com/ballerina-platform/ballerina-lang/issues/31671

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #34966

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by gimantha** [comment]
We need to change the base @pcnfernando

File: tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-negative.bal [hunk 1]
```diff
@@ -380,3 +380,11 @@ function testAssignabilityOfStreams() {
 function testNonIteratorsInStreamConstructor() {
     stream<int> intStrm = new("test");
 }
+
+function testUnboundedStreamWithEmptyConstructor() {
+    stream<int, error?> _ = new;
+    stream<int, error?> _ = new stream<int, error?>();
+    stream<int, readonly> _ = new stream<int, readonly>();
+    var _ = new stream<int, error?>();
```
Comment (lasinicl) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-negative.bal:388]: These are positive tests? Shall we move them to the respective file?
But `stream<int, readonly> _ = new stream<int, readonly>();` causes a TypeCastError at runtime?

Shall we also add some positive tests where the stream constructor is provided for the unbounded streams?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -402,6 +402,20 @@ public boolean containsErrorType(BType bType) {
         return type.tag == TypeTags.ERROR;
     }
 
+    public boolean containsNilType(BType bType) {
+        BType type = getReferredType(bType);
+        if (type.tag == TypeTags.UNION) {
+            return ((BUnionType) type).getMemberTypes().stream()
```
Comment (lasinicl) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:408]: Should we avoid using streams? :thinking:

Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Updated

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/stream/BStreamValueTest.java [hunk 3]
```diff
@@ -248,6 +253,12 @@ public void testStreamTypeNegative() {
         BAssertUtil.validateError(negativeResult, i++, "invalid stream constructor. expected a subtype " +
                 "of 'object { public isolated function next() returns record {| int value; |}?; }', but found " +
                 "'string'", 381, 31);
+//        todo should be enabled once #35847 is fixed
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/stream/BStreamValueTest.java:256]: Relates to https://github.com/ballerina-platform/ballerina-lang/issues/35847

File: tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-negative.bal (hunk 1 above)
Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-negative.bal:388]: Added the tests. readonly assignability has an issue https://github.com/ballerina-platform/ballerina-lang/issues/35847. Will be restricted at compile time after fixing it

**Review by gimantha — DISMISSED** [review]

**Review by gimantha — APPROVED** [review]

**Pull request merged**